experimental/algorithm/LAGraph_CFL_reachability.c

```diff
@@ -0,0 +1,324 @@
+    rule_error_s invalid_err = {0};
+    for (size_t i = 0; i < rules_count; i++) {
+        LAGraph_rule_WCNF rule = rules[i];
+
+        bool is_rule_eps = rule.prod_A == -1 && rule.prod_B == -1;
+        bool is_rule_term = rule.prod_A != -1 && rule.prod_B == -1;
+        bool is_rule_bin = rule.prod_A != -1 && rule.prod_B != -1;
+
+        // Check that all rules are well-formed
+        if (rule.nonterm < 0 || rule.nonterm >= nonterms_count) {
+            ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
+        }
+
+        // [Variable -> eps]
+        if (is_rule_eps) {
+            eps_rules[eps_rules_count++] = i;
+
+            continue;
+        }
+
+        // [Variable -> term]
+        if (is_rule_term) {
+            term_rules[term_rules_count++] = i;
+
+            if (rule.prod_A < -1 || rule.prod_A >= terms_count) {
+                ADD_INDEX_TO_ERROR_RULE(term_err, i);
+            }
+
+            continue;
+        }
+
+        // [Variable -> A B]
+        if (is_rule_bin) {
+            bin_rules[bin_rules_count++] = i;
+
+            if (rule.prod_A < -1 || rule.prod_A >= nonterms_count || rule.prod_B < -1 ||
+                rule.prod_B >= nonterms_count) {
+                ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
+            }
+
+            continue;
+        }
+
+        // [Variable -> _ B]
+        ADD_INDEX_TO_ERROR_RULE(invalid_err, i);
+    }
+
+    if (term_err.count + nonterm_err.count + invalid_err.count > 0) {
+        ADD_TO_MSG("Count of invalid rules: %ld.\n",
+                   term_err.count + nonterm_err.count + invalid_err.count);
+
+        if (nonterm_err.count > 0) {
+            ADD_TO_MSG("Non-terminals must be in range [0, nonterms_count). ");
+            ADD_TO_MSG("Indexes of invalid rules: %s\n", nonterm_err.indexes_str)
+        }
+        if (term_err.count > 0) {
+            ADD_TO_MSG("Terminals must be in range [-1, nonterms_count). ");
+            ADD_TO_MSG("Indexes of invalid rules: %s\n", term_err.indexes_str)
+        }
+        if (invalid_err.count > 0) {
+            ADD_TO_MSG("[Variable -> _ B] type of rule is not acceptable. ");
+            ADD_TO_MSG("Indexes of invalid rules: %s\n", invalid_err.indexes_str)
+        }
+
+        LG_FREE_ALL;
+        return GrB_INVALID_VALUE;
+    }
+
+    // Rule [Variable -> term]
+    for (size_t i = 0; i < term_rules_count; i++) {
+        LAGraph_rule_WCNF term_rule = rules[term_rules[i]];
+
+        GrB_Index nnz;
+        GrB_Matrix_nvals(&nnz, adj_matrices[term_rule.prod_A]);
+
+        GrB_Index row[nnz], col[nnz];
+        bool values[nnz];
+
+        GrB_Matrix_extractTuples(row, col, values, &nnz, adj_matrices[term_rule.prod_A]);
+        for (size_t j = 0; j < nnz; j++) {
+#ifdef DEBUG
+            printf("[TERM] SET ELEMENT [TRUE], NONTERM: %d, ROW: %ld, COL: %ld\n",
+                   term_rule.nonterm, row[j], col[j]);
+#endif
+            GrB_Matrix_setElement(T[term_rule.nonterm], true, row[j], col[j]);
+        }
+    }
```

Comment:
Is this method better than using `GrB_reduce` (_GrB_Matrix_reduce_Monoid_) + `GxB_Matrix_diag`?